// blueboat/build/main/lib/utils/Callback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Callback {
    constructor() {
        this.callbacks = [];
        this.removeCallback = (id) => {
            this.callbacks = this.callbacks.filter(callback => callback.id !== id);
        };
    }
    add(callback, onlyCallOnce) {
        const id = Math.random().toString();
        this.callbacks.push({
            callback,
            timesCalled: 0,
            canCallMultipleTimes: !onlyCallOnce,
            id
        });
        return { clear: () => this.removeCallback(id) };
    }
    clear() {
        this.callbacks.splice(0, this.callbacks.length);
    }
    call(argument, argument2) {
        this.callbacks = this.callbacks.map(callback => {
            if (callback.timesCalled > 0) {
                if (!callback.canCallMultipleTimes) {
                    return callback;
                }
            }
            callback.callback(argument, argument2);
            return Object.assign({}, callback, { timesCalled: callback.timesCalled + 1 });
        });
    }
}
exports.default = Callback;
