// blueboat/build/main/lib/server/CustomGameValues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RedisKeys_1 = require("../constants/RedisKeys");
/**
 * Used for global game values in which you can change in the admin panel
 */
class CustomGameValues {
    constructor(options) {
        this.storage = null;
        this.setGlobalGameValuesObject = async (newGameValues) => {
            try {
                await this.storage.set(RedisKeys_1.GAME_VALUES, JSON.stringify(newGameValues), {
                    noExpiration: true
                });
            }
            catch (e) {
                throw e;
            }
        };
        this.getGameValues = async () => {
            try {
                const gameValues = await this.storage.get(RedisKeys_1.GAME_VALUES, true);
                if (gameValues) {
                    return JSON.parse(gameValues);
                }
                return {};
            }
            catch (e) {
                return {};
            }
        };
        this.getGameValue = async (key, defaultValue) => {
            try {
                const gameValues = await this.getGameValues();
                if (gameValues[key]) {
                    return gameValues[key];
                }
                return defaultValue;
            }
            catch (e) {
                throw defaultValue;
            }
        };
        this.setGameValue = async (key, value) => {
            try {
                const gameValues = await this.getGameValues();
                const newGameValues = Object.assign({}, gameValues, { [key]: value });
                await this.storage.set(RedisKeys_1.GAME_VALUES, JSON.stringify(newGameValues), true);
            }
            catch (e) {
                throw e;
            }
        };
        this.resetGameValues = async () => {
            try {
                await this.storage.set(RedisKeys_1.GAME_VALUES, JSON.stringify({}), true);
            }
            catch (e) {
                throw e;
            }
        };
        this.storage = options.storage;
    }
}
exports.default = CustomGameValues;
