// blueboat/build/main/lib/server/RoomFetcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RedisKeys_1 = require("../constants/RedisKeys");
/**
 * Can help find a list of currently available Rooms and their snapshots
 */
class RoomFetcher {
    constructor(options) {
        this.storage = null;
        this.getListOfRooms = async () => {
            try {
                const rooms = await this.storage.fetchKeys(RedisKeys_1.ROOM_PREFIX);
                return rooms;
            }
            catch (e) {
                throw e;
            }
        };
        this.getListOfRoomsWithData = async () => {
            try {
                const roomList = await this.getListOfRooms();
                const rooms = await Promise.all(roomList.map(async (r) => {
                    try {
                        const room = await this.storage.get(RedisKeys_1.ROOM_PREFIX + r, true);
                        if (room) {
                            return JSON.parse(room);
                        }
                        return null;
                    }
                    catch (e) {
                        throw e;
                    }
                }));
                return rooms.filter(room => room !== null);
            }
            catch (e) {
                throw e;
            }
        };
        this.findRoomById = async (roomId) => {
            try {
                const room = await this.storage.get(RedisKeys_1.ROOM_PREFIX + roomId);
                return JSON.parse(room);
            }
            catch (e) {
                throw e;
            }
        };
        this.setRoomMetadata = async (roomId, newMetadata) => {
            try {
                const room = await this.findRoomById(roomId);
                await this.storage.set(RedisKeys_1.ROOM_PREFIX + room.id, JSON.stringify(Object.assign({}, room, { metadata: newMetadata })));
            }
            catch (e) {
                return e;
            }
        };
        this.addRoom = async (room) => {
            try {
                await this.storage.set(RedisKeys_1.ROOM_PREFIX + room.id, JSON.stringify(room));
            }
            catch (e) {
                throw e;
            }
        };
        this.removeRoom = async (roomId) => {
            try {
                await this.storage.remove(RedisKeys_1.ROOM_PREFIX + roomId);
            }
            catch (e) {
                throw e;
            }
        };
        this.storage = options.storage;
    }
}
exports.default = RoomFetcher;
