// blueboat/build/main/lib/server/Connection/ConnectionHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const nanoid_1 = __importDefault(require("nanoid"));
const serialize_error_1 = __importDefault(require("serialize-error"));
const ClientActions_1 = __importDefault(require("../../constants/ClientActions"));
const PubSubListeners_1 = require("../../constants/PubSubListeners");
const ServerActions_1 = __importDefault(require("../../constants/ServerActions"));
const Logger_1 = __importStar(require("../../utils/Logger"));
const CreateNewRoom_1 = __importDefault(require("./CreateNewRoom"));
const JoinRoom_1 = __importDefault(require("./JoinRoom"));
const ConnectionHandler = (options) => {
    const { io, socket, storage, pubsub, availableRoomTypes, roomFetcher, gameValues, onRoomMade, onRoomDisposed, customRoomIdGenerator } = options;
    const userId = socket.handshake.query.id || nanoid_1.default();
    const client = {
        id: userId,
        sessionId: socket.id,
        origin: socket &&
            socket.request &&
            socket.request.headers &&
            socket.request.headers.origin
            ? socket.request.headers.origin
            : '',
        ip: socket && socket.handshake && socket.handshake.address
            ? socket.handshake.address
            : ''
    };
    socket.emit(ServerActions_1.default.clientIdSet, userId);
    socket.on(ClientActions_1.default.createNewRoom, async (request) => {
        try {
            if (!request || !request.type || !request.uniqueRequestId) {
                throw new Error('Room type needed');
            }
            Logger_1.default(`${socket.id} trying to create a new ${request.type} room`, Logger_1.LoggerTypes.io);
            const room = await CreateNewRoom_1.default(client, io, roomFetcher, gameValues, pubsub, storage, availableRoomTypes, onRoomDisposed, request.type, await roomFetcher.getListOfRooms(), request.options, customRoomIdGenerator);
            Logger_1.default(`${room.roomId} made`, Logger_1.LoggerTypes.room);
            onRoomMade(room);
            socket.emit(`${request.uniqueRequestId}-create`, room.roomId);
        }
        catch (e) {
            const error = serialize_error_1.default(e);
            Logger_1.default(`${socket.id} error creating room - ${JSON.stringify(error)}`, Logger_1.LoggerTypes.room);
            socket.emit(`${request.uniqueRequestId}-error`, error);
        }
    });
    socket.on(ClientActions_1.default.joinRoom, async (payload) => {
        try {
            const { roomId } = payload;
            if (!roomId) {
                throw new Error('Room ID not provided');
            }
            Logger_1.default(`${socket.id} trying to join ${roomId} room`, Logger_1.LoggerTypes.io);
            await JoinRoom_1.default(roomId, client, roomFetcher, pubsub, payload.options);
            Logger_1.default(`${socket.id} joined ${roomId} room`, Logger_1.LoggerTypes.io);
        }
        catch (e) {
            if (payload && payload.roomId) {
                const error = serialize_error_1.default(e);
                Logger_1.default(`${socket.id} error joining room ${payload.roomId} - ${JSON.stringify(error)}`, Logger_1.LoggerTypes.room);
                socket.emit(`${payload.roomId}-error`, error);
            }
        }
    });
    socket.on(ClientActions_1.default.sendMessage, (message) => {
        if (message.key === undefined || !message.room) {
            return;
        }
        Logger_1.default(`${socket.id} - message - ${JSON.stringify(message)}`, Logger_1.LoggerTypes.io);
        pubsub.publish(message.room, {
            client,
            action: ClientActions_1.default.sendMessage,
            data: { key: message.key, data: message.data }
        });
    });
    socket.on('disconnect', () => {
        Logger_1.default(`${socket.id} - disconnected`, Logger_1.LoggerTypes.io);
        pubsub.publish(PubSubListeners_1.PLAYER_LEFT, socket.id);
    });
};
exports.default = ConnectionHandler;
