// blueboat/build/main/lib/server/Connection/CreateNewRoom.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const nanoid_1 = __importDefault(require("nanoid"));
const CreateNewRoom = (client, io, roomFetcher, gameValues, pubsub, storage, availableRooms, onRoomDisposed, roomName, existingRoomIds, creatorOptions, customRoomIdGenerator) => {
    return new Promise(async (resolve, reject) => {
        try {
            const roomToCreate = availableRooms.filter(r => r.name === roomName)[0];
            if (!roomToCreate) {
                throw new Error(`${roomName} does not have a room handler`);
            }
            let roomId;
            for (let i = 0; i < 3; i++) {
                if (roomId) {
                    break;
                }
                const possibleRoomId = customRoomIdGenerator
                    ? customRoomIdGenerator(roomToCreate.name, roomToCreate.options, creatorOptions)
                    : nanoid_1.default();
                if (!existingRoomIds.includes(possibleRoomId)) {
                    roomId = possibleRoomId;
                }
            }
            if (!roomId) {
                throw new Error('Failed to create room with unique ID');
            }
            const initialGameValues = await gameValues.getGameValues();
            const room = new roomToCreate.handler({
                io,
                pubsub,
                owner: client,
                roomId,
                storage,
                creatorOptions,
                options: roomToCreate.options,
                roomFetcher,
                gameValues,
                initialGameValues,
                onRoomDisposed,
                roomType: roomToCreate.name,
                onRoomCreated: (error) => {
                    if (!error) {
                        const snapshot = {
                            id: roomId,
                            type: roomName,
                            owner: client,
                            metadata: {},
                            createdAt: Date.now()
                        };
                        roomFetcher
                            .addRoom(snapshot)
                            .then(() => resolve(room))
                            .catch(e => reject(e));
                    }
                    else {
                        reject(error);
                    }
                }
            });
        }
        catch (e) {
            reject(e);
        }
    });
};
exports.default = CreateNewRoom;
