// blueboat/build/main/example/ChatRoom.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../index");
const State_1 = __importDefault(require("./State"));
class ChatRoom extends index_1.Room {
    constructor() {
        super(...arguments);
        this.onLeave = async (client) => {
            const reconnected = await this.allowReconnection(client, 30);
            if (reconnected) {
                return;
            }
            this.addMessage(`${client.id} has left the room`, 'Botsy');
        };
        this.addMessage = (text, senderId) => {
            const message = { message: text, senderId };
            this.state.messages.push(message);
            this.broadcast('MESSAGE', message);
        };
        this.catchUp = (client) => client.send('MESSAGES', this.state.messages);
    }
    async onCreate() {
        try {
            this.setState(new State_1.default(this.initialGameValues.initialBotMessage || 'Welcome to the chat!'));
        }
        catch (e) {
            throw e;
        }
    }
    onJoin(client) {
        this.catchUp(client);
    }
    onMessage(client, action, data) {
        if (!action) {
            return;
        }
        if (action === 'LATENCY') {
            client.send('LATENCY', data);
        }
        if (action === 'CHAT') {
            this.addMessage(data, client.id);
        }
    }
    async beforeDispose() {
        this.broadcast('DISPOSED');
    }
}
exports.default = ChatRoom;
