// blueboat/build/main/example/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cors_1 = __importDefault(require("cors"));
const express_1 = __importDefault(require("express"));
const index_1 = require("../index");
const ChatRoom_1 = __importDefault(require("./ChatRoom"));
const redisOptions = {
    host: 'localhost',
    port: 6379
};
const start = async () => {
    try {
        const app = express_1.default();
        app.use(cors_1.default());
        app.options('*', cors_1.default());
        const server = new index_1.Server({
            app,
            storage: index_1.ClusterMemoryStorage(),
            pubsub: index_1.EventEmitterClusterPubsub.PubSub(),
            adapters: [index_1.EventEmitterClusterPubsub.ClusterAdapter()],
            redis: redisOptions,
            admins: { blueboat: 'pass' }
        });
        server.registerRoom('Chat', ChatRoom_1.default);
        server.listen(4000, () => {
            console.log('Server listening on port 4000');
        });
    }
    catch (e) {
        console.log(e);
    }
};
index_1.EventEmitterClusterPubsub.ProcessStarter(start, 2);
